### DISCUSSION TITLE
Trying to use Viomi S9 vacuum but have issues

### DISCUSSION THREAD

**Discussion by fokcuk** [rytilahti/python-miio#1081] [Q&A]
Hi
I am trying to do some basic cli testing of commands as I am not that good at Python to write code from scratch.
I have a Viomi S9 vacuum. When I try to run this command I get an error about **hw_ver**:
─# miiocli viomivacuum --ip 192.168.1.124 --token 78326372527131116c6a436a4a6c3744 info                                                                 2 ⨯
Traceback (most recent call last):
  File "/usr/local/bin/miiocli", line 8, in <module>
    sys.exit(create_cli())
  File "/usr/local/lib/python3.9/dist-packages/miio/cli.py", line 63, in create_cli
    return cli(auto_envvar_prefix="MIIO")
  File "/usr/local/lib/python3.9/dist-packages/miio/click_common.py", line 59, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python3/dist-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/lib/python3/dist-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/lib/python3/dist-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/lib/python3/dist-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/miio/click_common.py", line 285, in wrap
    result_msg = result_msg_fmt.format(**kwargs)
  **File "/usr/local/lib/python3.9/dist-packages/miio/device.py", line 88, in hardware_version
    if self.data["hw_ver"] is not None:
KeyError: 'hw_ver'**

I have seen this error in another project that is connecting to Xiaomi devices. What I ended up doing there was to modify device.py and comment out all lines that tried to check for **hardware_version**. 

Just trying to get to it working one line at a time and would appreciate some help

**Comment by fokcuk** [comment]
found more info =) but stuck at a stupid error
─# python3 miottemplate.py download urn:miot-spec-v2:device:vacuum:0000A006:viomi-v18:1    
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/dist-packages/miio/miottemplate.py", line 6, in <module>
    from containers import Device, ModelMapping
ModuleNotFoundError: No module named 'containers'

**Reply by rytilahti** [reply]
`miottemplate.py` should not be installed at all, so that's odd... You should execute the command inside the `devtools` directory and it should work then!

**Reply by fokcuk** [reply]
ok. I will try. Can you please tell me what command to use to call "actions" and not get_properties?

**Reply by rytilahti** [reply]
The API (nor the documentation) is great on that part, but if you create mapping like in the `_miot.py` files containing an action (i.e., an entry containing `siid` and `aiid`), you can use `call_action(name)` to call it.

If you simply want to call a specific `siid`&`aiid` combination, the you can also use `call_action_by(siid, aiid)` (https://github.com/rytilahti/python-miio/blob/master/miio/miot_device.py#L90) to do just that.

**Reply by fokcuk** [reply]
is that for the use with raw_command or in place?

**Reply by fokcuk** [reply]
Just tried - did not work
└─# miiocli device --ip 192.168.1.124 --token 78326372527134446c4a436a1a6c3714 raw_command call_action_by "[{'did': 'MYDID', 'siid': 2, 'aiid': 1 }]"
Running command raw_command
Error: {'code': -9999, 'message': 'user ack timeout'}

**Reply by fokcuk** [reply]
cannot work out how to get it to process any action commands =(

**Reply by rytilahti** [reply]
Oh, you didn't want to use the API but the command line. There is a bug in miotdevice handling making it not to work currently (#1085 - if you comment out those checks, you should be able to use `miiocli miot_device call_action_by <siid> <aiid> <data>`) but you could also try:
```
$ miiocli device raw_command action '{"did": "something", "siid": 2, "aiid": 1, "in": []}'
```

See the link I posted above, it shows the payload keys (and I think the `in` was required, even if you don't want to pass parameters to the call).

**Reply by fokcuk** [reply]
yep! that worked!


I am using command line as I am trying to work out if zone/room cleaning will work at all. I am not that python literate to use API at the moment. Once I sort it out, I will reuse the cade to create another function - if it works

**Reply by fokcuk** [reply]
can you do multiple siid/piid changes in the same set_attribues?

**Reply by fokcuk** [reply]
and, if you be so kind =) when you have something like this:

siid 4: aiid 13 Choose room to clean: in: [36, 37, 38] -> out: []

What do you need to pass as an "in" values when calling the action?

**Reply by rytilahti** [reply]
> can you do multiple siid/piid changes in the same set_attribues?

The API does not implement this, but you can try to pass an array of objects to the `set_properties` command to see if your device accepts that.

> What do you need to pass as an "in" values when calling the action?

Unfortunately I don't know an answer for that. I think I read somewhere else that the room numbers are set by the device/the app, but I'm not sure about the expected format.

**Reply by fokcuk** [reply]
the reason I ask is because
 clean: in: [36, 37, 38] -> out: []
contains piid write only attributes 36,37,38. So was asking if you how to work with those - not room specific

**Reply by rytilahti** [reply]
Oh, so the separate "attribute" IDs are called `piid` (for property id), so the syntax you see here should work: https://github.com/rytilahti/python-miio/blob/master/miio/miot_device.py#L136 - the `value` contains the value to be set to the given property.

For some properties the json file (or the output of miottemplate) may give some hints about the type of the value, sometimes also things like accepted ranges or values (which can also be potentially induced in cases where there exists a readable property).

Hope this helps!

**Reply by fokcuk** [reply]
yeah, setting one attribute is fine. I just dont know how the action works when it has references to the attribute ids.
I tried calling it like that:

miiocli device raw_command action '{"did": "something", "siid": 2, "aiid": 1, "in": [0, 1, 8]}' 

but that did not work

This is how its defined in the urn:

Property:
siid 4: piid: 36 (clean-room-mode): (uint8, unit: none) (acc: ['write'])
{'value': 0,'description':'Global'}
{'value': 1,'description':'Along the edge'}
siid 4: piid: 37 (clean-room-oper): (uint8, unit: none) (acc: ['write'])
{'value': 0,'description':'stop'}
{'value': 1,'description':'Start'}
{'value': 2,'description':'Pause'}
siid 4: piid: 38 (clean-room-ids): (string, unit: none) (acc: ['write'])

Action:
siid 4: aiid 13 Choose room to clean: in: [36, 37, 38] -> out: []

**Reply by fokcuk** [reply]
sorry, my bad - it did work on another action. The one I want to run it on did not work 
Running command raw_command
Error: {'code': -9999, 'message': 'user ack timeout'}
